pkg/receive/handler.go

```diff
@@ -964,20 +967,34 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 		return errors.New("hashring is not ready")
 	}
 
-	for i = 0; i < h.options.ReplicationFactor; i++ {
-		endpoint, err := h.hashring.GetN(tenant, &wreq.Timeseries[0], i)
-		if err != nil {
-			h.mtx.RUnlock()
-			return err
+	replicatedRequests := make(map[endpointReplica]*prompb.WriteRequest)
```

Comment:
Might impact benchmarks as it's a hot path 🤔 Worth to be careful here, but yes we have to now.

Reply:
I wonder if we should expand hashring interface or pass this flag outside and only do that for ketama then?

Reply:
I had the same dilemma and not sure what's better. I'll write some benchmarks to see whether optimizing it makes sense.

Reply:
Yeah, it's a bit unfortunate as it seems we're effectively 'rebatching' the time series for the replication requests. I was wondering if we'd reshuffled a bit the logic and did the replication batching in one sweep in the `forward` method? If we know we are going to replicate the request, we could batch the replication requests and re-use the map we have in the `forward`. Could that save us something in performance?

Reply:
I think that's a good idea, I'll give it a try

Reply:
Seems that making further optimizations does not make sense at this point. For the `hashmod` ring, we still create the same amount of requests as before because each endpoint will have exactly one replica. So even though the key is `replica-endpoint`, instead of just `endpoint`, the number of these keys will be same as before.

For the `ketama` hashring we cannot avoid having multi-replica endpoints. 
Benchmarks also show no significant change, which is likely because most of the latency comes from marshaling and sending data over the wire, as well as appending to TSDB. Hashing seems to be quite cheap in this context.


```
benchstat  .bench/main .bench/branch 
name                                                                               old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-8                         534µs ± 0%     978µs ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-8            615µs ± 0%     669µs ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-8                       5.34ms ± 0%    5.76ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-8          6.10ms ± 0%    6.37ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-8                 53.6ms ± 0%    51.3ms ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-8    52.1ms ± 0%    51.1ms ± 0%   ~     (p=1.000 n=1+1)

name                                                                               old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-8                        1.20MB ± 0%    1.20MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-8           1.46MB ± 0%    1.46MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-8                       13.5MB ± 0%    13.5MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-8          16.0MB ± 0%    16.0MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-8                  120MB ± 0%     120MB ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-8     120MB ± 0%     120MB ± 0%   ~     (p=1.000 n=1+1)

name                                                                               old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/OK-8                         3.10k ± 0%     3.10k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_500_of_them/conflict_errors-8            6.64k ± 0%     6.63k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/OK-8                        30.3k ± 0%     30.3k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/typical_labels_under_1KB,_5000_of_them/conflict_errors-8           65.2k ± 0%     65.1k ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/OK-8                   90.0 ± 0%      87.0 ± 0%   ~     (p=1.000 n=1+1)
HandlerReceiveHTTP/extremely_large_label_value_10MB,_10_of_them/conflict_errors-8       219 ± 0%       218 ± 0%   ~     (p=1.000 n=1+1)
```

Reply:
Great!  💪🏽